hypothesis-python/src/hypothesis/extra/gufunc.py

```diff
@@ -0,0 +1,451 @@
+DEFAULT_MAX_SIDE = 5
+
+# This uses "private" function of numpy, but it does the job. It throws a
+# pretty readable exception for invalid input, so we don't need to add anything
+# there. We should also check signature.isascii() since there are lot of weird
+# corner cases with unicode parsing, but isascii() restricts us to Py >=3.7.
+parse_gufunc_signature = npfb._parse_gufunc_signature
+
+
+def check_set_like(arg, name=""):
+    """Validate input can be searched like a `set`."""
+    try:
+        0 in arg
+    except TypeError:
+        raise InvalidArgument("Expected set-like but got %s=%r (type=%s)"
+                              % (name, arg, type(arg).__name__))
+
+
+def check_valid_size_interval(min_size, max_size, name, floor=0):
+    """Check valid for integers strategy and array shapes."""
+    # same checks as done in integers
+    check_valid_bound(min_size, name)
+    check_valid_bound(max_size, name)
+    order_check(name, floor, min_size, max_size)  # ensure non-none & above 0
+    # this is also done in integers, so check for good measure
+    check_valid_interval(min_size, max_size, 'min_size', 'max_size')
+
+
+def order_check_min_max(min_dict, max_dict):
+    """Check min and max dict compatible with integers and array shapes."""
+    check_valid_size_interval(min_dict.default_factory(),
+                              max_dict.default_factory(), "side default")
+    for kk in (set(min_dict.keys()) | set(max_dict.keys())):
+        check_valid_size_interval(min_dict[kk], max_dict[kk], "side %s" % kk)
+
+
+def ensure_int(arg, name=""):
+    """Validate input as `int` and return it."""
+    try:
+        x = int(arg)
+        assert arg == x  # e.g., check 5.0 and not 5.5 was passed
+    except Exception:
+        raise InvalidArgument("%s=%r (type=%s) not representable as int"
+                              % (name, arg, type(arg).__name__))
+    return x
+
+
+def _int_or_dict(x, default_val):
+    """Pre-process cases where argument `x` can be `int`, `dict`, or
+    `defaultdict`. In all cases, build a `defaultdict` and return it.
+    """
+    # case 1: x already defaultdict, leave it be, pass thru
+    if isinstance(x, defaultdict):
+        return x
+
+    default_val = ensure_int(default_val, "default value")
+    try:
+        # case 2: x is or can be converted to dict
+        D = defaultdict(lambda: default_val, x)
+    except TypeError:
+        # case 3: x is or can be converted to int => make a const dict
+        default_val = ensure_int(x, "constant value")
+        D = defaultdict(lambda: default_val)
+    # case 4: if can't be converted to dict or int, then exception raised
+    return D
+
+
+@composite
+def _arrays(draw, dtype, shape, elements=None, unique=False):
+    """Wrapper to fix issues with `hypothesis.extra.numpy.arrays`.
+
+    `arrays` is strict on shape being `int` which this fixes. This is partially
+    not needed in Py3 since there is no `int` vs `long` issue. Also, `arrays`
```

Comment:
Whoops!  I'd consider that a bug in the current `arrays` strategy; it should be checking against `compat.integer_types` rather than `int`.

May I suggest opening a smaller `patch` PR to fix that?

Then you'll be able to delete this wrapper :smile:

Reply:
This wrapper serves two purposes:
1) `arrays` does not accept `long` (or numpy types)
2) It misbehaves for 0-dimensional arrays:
```
>>> from hypothesis.extra.numpy import arrays
>>> import random
>>> random.seed(0)
>>> S=arrays('>i2',())
>>> x=S.example()
>>> x
-31543
>>> x.dtype
dtype('int16')
>>> x.dtype == '>i2'
False
>>> type(x)
<class 'numpy.int16'>
```
`x` is not an array and also its dtype is not the requested endianess. This only seems to happen for 0-dimensional arrays.

I can see how to fix the `long` issue, but I can't find the source of the 2nd issue.

Reply:
https://github.com/HypothesisWorks/hypothesis/blob/d879f8277ed7398ec4e45d56d50cc55d42e7161b/hypothesis-python/src/hypothesis/extra/numpy.py#L395-L397

Looks like we explicitly handle zero-dimensional shapes by returning a scalar - if you want to change this I'd be happy to discuss that as a separate PR just to keep the scope manageable.

Reply:
So, would the new interface be to add an optional argument `always_array=False`?

Reply:
It definitely looks like the handling of the scalar case isn't as intended, since it doesn't even ensure the specified type as in the non-scalar case:
```
>>> from hypothesis.strategies import booleans
>>> from hypothesis.extra.numpy import arrays
>>> import random
>>> random.seed(0)
>>> S=arrays('int32',(),elements=booleans())
>>> x=S.example()
>>> x
True
>>> type(x)
<class 'bool'>
>>> S=arrays('int32',(1,),elements=booleans())
>>> x=S.example()
>>> x
array([1], dtype=int32)
>>> x.dtype
dtype('int32')
```
Perhaps lines 395--397 were placed there for efficiency (avoid creating an `ArrayStrategy` object) rather than desired functionality?

Reply:
Pretty sure it's just there for efficiency - but also because the `ArrayStrategy` internals assume the array is at least one-dimensional.  What if we replace

https://github.com/HypothesisWorks/hypothesis/blob/d879f8277ed7398ec4e45d56d50cc55d42e7161b/hypothesis-python/src/hypothesis/extra/numpy.py#L395-L397

with 

```python
if not shape:
	return np.full(shape=(), fill_value=draw(elements), dtype=dtype)
```

If that works, it should be a separate PR before this one is merged.

Reply:
Or replace `test_gen_data.py.test_empty_dimensions_are_scalars` with `empty_dimensions_are_arrays`!

Reply:
Yeah, I think switching to testing `empty_dimensions_are_arrays` would better if the backwards compatibility implications are not a problem.

Reply:
IMO the current behaviour is a design bug, which is also a rare edge case. I'd suggest opening a PR for a minor release, and we'll seek comment from David and Ryan.